### PR TITLE
Baccarat variants

### PR DESCRIPTION
**Pull request by mgirman520**
Ace of Swords and Ace of Sorrows

### CODE REVIEW

File: Controller/Heroes/Baccarat/CharacterCards/AceOfSwordsSwordsBaccaratCharacterCardController.cs [hunk 1]
```diff
@@ -0,0 +1,158 @@
...
+                coroutine = base.GameController.SelectAndMoveCard(base.HeroTurnTakerController, (Card c) => c.Identifier == "AfterlifeEuchre" && !playedCards.Contains(c) && c.Location == base.TurnTaker.Trash, base.TurnTaker.PlayArea, optional: true, playIfMovingToPlayArea: true, storedResults: decisions, cardSource: base.GetCardSource());
+                if (base.UseUnityCoroutines)
+                {
+                    yield return base.GameController.StartCoroutine(coroutine);
+                }
+                else
+                {
+                    base.GameController.ExhaustCoroutine(coroutine);
+                }
+                selectedCard = decisions.LastOrDefault<SelectCardDecision>().SelectedCard;
```
Comment (Vilnor) [inline Controller/Heroes/Baccarat/CharacterCards/AceOfSwordsSwordsBaccaratCharacterCardController.cs:134]: Prefer DidSelectCard() and GetSelectedCard() built-ins

File: Controller/Heroes/Baccarat/CharacterCards/AceOfSwordsSwordsBaccaratCharacterCardController.cs [hunk 2]
```diff
@@ -0,0 +1,158 @@
...
+                        {
+                            base.GameController.ExhaustCoroutine(coroutine2);
+                        }
+                        break;
+                    }
+                case 2:
+                    {
+                        //Each hero character may deal themselves 1 toxic damage to draw a card now.
+                        this.actedHeroes = new List<Card>();
+                        IEnumerable<Function> functionsBasedOnCard(Card c) => new Function[]
```
Comment (Vilnor) [inline Controller/Heroes/Baccarat/CharacterCards/AceOfSwordsSwordsBaccaratCharacterCardController.cs:61]: I'm not sure you need this private variable if you instead use SelectTurnTakersAndDoAction.  That will also give you a place to select which of the turn taker's character cards take the damage if that's required.  There's also YesNoAction_ManyPlayers that looks even better.

File: Controller/Heroes/Baccarat/CharacterCards/AceOfSwordsSwordsBaccaratCharacterCardController.cs [hunk 3]
```diff
@@ -0,0 +1,158 @@
...
+                    }
+            }
+            yield break;
+        }
+        private IEnumerator DealDamageAndDrawResponse(Card card)
+        {
+            if (card != null)
+            {
+                IEnumerator coroutine = base.DealDamage(card, card, 1, DamageType.Toxic, false, false, false, null, null, null, false, null);
+                IEnumerator coroutine2 = base.DrawCard(card.Owner.ToHero());
```
Comment (Vilnor) [inline Controller/Heroes/Baccarat/CharacterCards/AceOfSwordsSwordsBaccaratCharacterCardController.cs:84]: Please remove Default arguments taht you aren't specifically setting.

**Review by Vilnor — CHANGES_REQUESTED** [review]
Base Card Tests should be parameterized to test both base and promo characters.
Tests that don't work on the promo (Innate power needing different choices) should be copied over to the PromoTests.

Specific concerns on testing are the extra Innate Powers card.

**Comment by Vilnor** [comment]
Note that I didn't read the Promo Tests in detail.

File: Controller/Heroes/Baccarat/CharacterCards/AceOfSwordsSwordsBaccaratCharacterCardController.cs (hunk 1 above)
Comment (mgirman520) [reply Controller/Heroes/Baccarat/CharacterCards/AceOfSwordsSwordsBaccaratCharacterCardController.cs:134]: DidSelectCard() returns false before first pass, causing the power to not trigger,
GetSelectedCard() returns the last selected card, it ignores skips/nulls

File: Controller/Heroes/Baccarat/CharacterCards/AceOfSwordsSwordsBaccaratCharacterCardController.cs (hunk 2 above)
Comment (mgirman520) [reply Controller/Heroes/Baccarat/CharacterCards/AceOfSwordsSwordsBaccaratCharacterCardController.cs:61]: This is how Scholar's Proverbs and Axioms implements it

File: Controller/Heroes/Baccarat/CharacterCards/AceOfSwordsSwordsBaccaratCharacterCardController.cs (hunk 3 above)
Comment (mgirman520) [reply Controller/Heroes/Baccarat/CharacterCards/AceOfSwordsSwordsBaccaratCharacterCardController.cs:84]: Removed.

**Comment by mgirman520** [comment]
> Base Card Tests should be parameterized to test both base and promo characters.
> Tests that don't work on the promo (Innate power needing different choices) should be copied over to the PromoTests.
> 
> Specific concerns on testing are the extra Innate Powers card.

Tested and confirm works as intended

**Comment by mgirman520** [comment]
Closing until variant loading is figured out

**Pull request closed without merging**